src/coreclr/jit/lower.cpp

```diff
@@ -3379,6 +3379,12 @@ GenTree* Lowering::LowerJTrue(GenTreeOp* jtrue)
             flags   = cond->OperIs(GT_EQ) ? GTF_JCMP_EQ : GTF_EMPTY;
             useJCMP = true;
         }
+        else if (cond->OperIs(GT_LT, GT_GE) && relopOp2->IsIntegralConst(0))
+        {
+            // Codegen will use tbnz or tbz in codegen which do not affect the flag register
+            flags   = cond->OperIs(GT_LT) ? GTF_JCMP_LT : GTF_JCMP_GE;
+            useJCMP = true;
+        }
```

Comment:
Seems like it would be simpler to phrase this in terms of the existing support for `JCMP` by changing the 0-constant into the upper bit and using `GTF_JCMP_EQ/GTF_EMPTY` like above. Then you wouldn't need to add new handling in the backend/more flags.

Reply:
More generally we could transform `x < 0` into `TEST_EQ(x, upper_bit)` in `OptimizeConstCompare`, and it would be picked up automatically here -- but maybe that's not beneficial when the compare is not being consumed by a `JTRUE`?

Reply:
Good points. I think I see how to use the existing flags.
We can just change the handling of JCMP_TST to handle the 0-constant in code-gen.

Reply:
I don't even think that is necessary -- you can change the constant right here, e.g. `relopOp2->AsIntConCommon()->SetIconValue(1 << (8 * genTypeSize(genActualType(relopOp1)) - 1))`

Reply:
For now, lets just handle the transformation here; it's not much code to handle it like this.

Reply:
I disagree -- this is changing the semantics of `JCMP`. Suddenly, its semantics relies on the value of an operand node. We should not have that kind of IR semantics.

Reply:
So what should I do exactly?

Reply:
I can make this change:
> (1 << (8 * genTypeSize(genActualType(relopOp1)) - 1))

when setting the icon value.

Reply:
Looks like we'll also need to change the assert in codegen to pass `size_t` to `isPow2` (the `genLog2` call below it already looks ok)

Reply:
I think parts of this would be simpler if we had a more specific op to handle this, like a `GT_JCMP_BIT_ZERO` so we can avoid having to do `(1 << (8 * genTypeSize(genActualType(relopOp1)) - 1))` to make it line up with the semantics of `GT_JCMP`.

@BruceForstall and I were talking about what if we had more machine specific ops to do what we want rather than rely on containment and having to make changes in code-gen to handle the contained nodes.

Reply:
I agree, it would make sense to me to have a node that directly represents `tbnz`/`tbz`, e.g. `JBT` (for jump on bit test). Having this weird `GTF_JCMP_TST` flag to alter the behavior of `JCMP` is pretty odd.

Reply:
Could even just be `GT_TBNZ`/`GT_TBZ` directly, we have some prior art with `GT_BT` and `GT_CCMP` for that kind of naming convention. I guess it would need a new `GenTree` kind though if we want to represent the bit position as an immediate instead of a contained constant value.